xsnippet_api/application.py

```diff
@@ -27,19 +30,76 @@ def create_app(conf):
     :return: an application instance
     :rtype: :class:`aiohttp.web.Application`
     """
-    router_v1 = web_urldispatcher.UrlDispatcher()
-    router_v1.add_route('*', '/snippets', resources.Snippets)
-    router_v1.add_route('*', '/snippets/{id}', resources.Snippet)
-
-    app = web.Application(router=router.VersionRouter(
-        {
-            '1': router_v1,
-        }
-    ))
+    # we need to import all the resources in order to evaluate @endpoint
+    # decorator, so they can be collected and passed to VersionRouter
+    from . import resources  # noqa
+    app = web.Application(router=router.VersionRouter(endpoint.collect()))
 
     # attach settings to the application instance in order to make them
     # accessible at any point of execution (e.g. request handling)
     app['conf'] = conf
     app['db'] = database.create_connection(conf)
 
     return app
+
+
+class endpoint:
+    """Define a RESTful API endpoint.
+
+    Usage example:
+
+        @endpoint('/myresources/{id}', '1.0')
+        class MyResource(web.View):
+            def get(self):
+                pass
+
+    :param route: a route to a wrapped resource
+    :param version: initial supported API version
+    :param end_version: last supported version; currently supported if None
+    """
+
+    _Item = collections.namedtuple('ResourceItem', [
+        'resource',
+        'route',
+        'version',
+        'end_version',
+    ])
+
+    _registry = []
+
+    def __init__(self, route, version, end_version=None):
+        self._options = (route, version, end_version)
+
+    def __call__(self, resource):
+        self._registry.append(self._Item(resource, *self._options))
+        return resource
+
+    @classmethod
+    def collect(cls):
+        rv = collections.defaultdict(web_urldispatcher.UrlDispatcher)
+
+        # Create routers for each discovered API version. The main reason why
+        # we need that so early is to register resources in all needed routers
+        # according to supported version range.
+        for item in cls._registry:
+            rv[item.version]
+
+        for item in cls._registry:
+            # if there's no env_version then a resource is still working, and
+            # latest discovered version should be considered as end_version
+            end_version = item.end_version or sorted(
+                rv.keys(),
+                key=lambda v: pkg_resources.parse_version(v)
```

Comment:
Yep, you're right. At the very beginning I've been using another API, and lambda was required. :)